### PR TITLE
[core]Update boost in ray deps

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?
Update boost version to the latest ones.

- updated the rules to build boost
- update the bazel version so it can work with the latest rules
- update boost version

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by scv119 — APPROVED** [review]

**Comment by iycheng** [comment]
@ericl  you are the code owner, could you take a look of this?

File: python/setup.py [hunk 1]
```diff
@@ -24,7 +24,7 @@
 # When the bazel version is updated, make sure to update it
 # in WORKSPACE file as well.
 
-SUPPORTED_BAZEL = (4, 2, 2)
+SUPPORTED_BAZEL = (5, 4, 0)
```
Comment (ericl) [inline python/setup.py:27]: This will require everyone to upgrade bazel. Just want to make sure we're ok with that (cc @rkooo567 @krfricke )

**Review by ericl — APPROVED** [review]

File: python/setup.py (hunk 1 above)
Comment (rkooo567) [reply python/setup.py:27]: Not aware of any particular issues

**Review by rkooo567 — APPROVED** [review]

**Comment by iycheng** [comment]
Can't reproduce the failure:
```
All wheels passed the sanity check and have the correct wheel commit set.
```

I think it's a transient issue and boost's upgrade + bazel upgrade just invalidate the cache. I'll merge this for now.

**Pull request merged**